File: SKIBin_old/app/src/main/java/com/h2Invent/skibin/ChildDetailActivity.java
package com.h2Invent.skibin;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;

import android.Manifest;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.CheckBox;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import static com.h2Invent.skibin.FragmentChildListCheckin.EXTRA_URL;

public class ChildDetailActivity extends AppCompatActivity {
    private JSONObject data;
    private RequestQueue requestQueue;
    private static String detailUrl;
    private TextView textViewName;
    private LinearLayout linearLayoutInfo;
    private LinearLayout linearLayoutinfoBool;
    private FloatingActionButton callerButton;
    private String phoneNumber;
    private String notfallKontakt;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_child_detail);

        callerButton = (FloatingActionButton) findViewById(R.id.callParents);
        callerButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                    call(phoneNumber,notfallKontakt);
            }
        });
        Intent intent = getIntent();
        detailUrl = intent.getStringExtra(EXTRA_URL);
        textViewName = (TextView) findViewById(R.id.childDetailName);
        linearLayoutInfo = (LinearLayout) findViewById(R.id.childDettailInfo);
        linearLayoutinfoBool = (LinearLayout) findViewById(R.id.childDetailBoolean);
        requestQueue = Volley.newRequestQueue(this);
        String uri = String.format(detailUrl + "?communicationToken=%1$s",
                ChildListMainActivity.userToken);
        JsonObjectRequest request = new JsonObjectRequest(Request.Method.GET, uri, null,
                new Response.Listener<JSONObject>() {
                    @Override
                    public void onResponse(JSONObject response) {
                        try {
                            if (response.getBoolean("error")) {
                                //todo errorText in eine Snackbar
                            } else {
                                String name = response.getString("vorname") + " " + response.getString("name");
                                textViewName.setText(name);
                                phoneNumber = response.getString("phone");
                                notfallKontakt = response.getString("parentsName");
                                JSONArray info = response.getJSONArray("info");
                                JSONArray infoBool = response.getJSONArray("boolean");
                                final TextView[] tv = new TextView[info.length() * 2 + 1];
                                for (int i = 0; i < info.length(); i++) {
                                    JSONObject inf = info.getJSONObject(i);
                                    TextView tvTmpInf = new TextView(getApplicationContext());

                                    String infName = inf.getString("name");
                                    String invValue = inf.getString("value") != "null" ? inf.getString("value") : "Keine Angabe";
                                    tvTmpInf.setText(invValue);
                                    tvTmpInf.setTextSize((float) 20);
                                    tvTmpInf.setPadding(0, 16, 20, 0);
                                    tvTmpInf.setTextColor(getApplicationContext().getResources().getColor(R.color.design_default_color_primary_dark));
                                    linearLayoutInfo.addView(tvTmpInf);
                                    TextView tvTmpLabel = new TextView(getApplicationContext());
                                    tvTmpLabel.setText(infName);
                                    tvTmpLabel.setTextSize((float) 12);
                                    tvTmpLabel.setPadding(0, -1, 20, 0);

                                    tvTmpLabel.setTextColor(getApplicationContext().getResources().getColor(R.color.grey));
                                    linearLayoutInfo.addView(tvTmpLabel);
                                }
                                for (int i = 0; i < infoBool.length(); i++) {
                                    JSONObject inf = infoBool.getJSONObject(i);
                                    CheckBox cb = new CheckBox(getApplicationContext());
                                    cb.setText(inf.getString("name"));

                                    cb.setEnabled(false);
                                    cb.setTextColor(getResources().getColor(R.color.black));

                                    if (inf.getBoolean("value")) {
                                       cb.setChecked(true);
                                    } else {
                                     cb.setChecked(false);
                                    }

                                    linearLayoutinfoBool.addView(cb);
                                }


                            }

                        } catch (JSONException e) {
                            e.printStackTrace();
                        }

                    }
                },
                new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        error.printStackTrace();
                    }
                });
        requestQueue.add(request);
    }

    public JSONObject getData() {
        return data;
    }

    public void setData(JSONObject data) {
        this.data = data;
    }


    private void call(final String phoneNumber, String parents){
        new AlertDialog.Builder(ChildDetailActivity.this)
                .setTitle(parents)
                .setMessage(R.string.callTitle)

                // Specifying a listener allows you to take an action before dismissing the dialog.
                // The dialog is automatically dismissed when a dialog button is clicked.
                .setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                        Intent callIntent = new Intent(Intent.ACTION_CALL);
                        callIntent.setData(Uri.parse("tel:" + phoneNumber));

                        if (ActivityCompat.checkSelfPermission(ChildDetailActivity.this,
                                Manifest.permission.CALL_PHONE) != PackageManager.PERMISSION_GRANTED) {

                            Toast.makeText(ChildDetailActivity.this, R.string.noCallPermission, Toast.LENGTH_SHORT).show();
                            return;
                        }
                        startActivity(callIntent);
                    }
                })

                // A null listener allows the button to dismiss the dialog and take no further action.
                .setNegativeButton(android.R.string.no, null)
                .show();
    }
}

File: app/src/main/java/com/h2Invent/skibin/ChildDetailActivity.kt
package com.h2Invent.skibin

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.android.volley.Request
import com.android.volley.toolbox.JsonObjectRequest
import com.android.volley.toolbox.Volley
import com.google.android.material.button.MaterialButton
import com.google.android.material.card.MaterialCardView
import org.json.JSONObject
import java.util.Locale

class ChildDetailActivity : AppCompatActivity() {
    private val requestQueue by lazy { Volley.newRequestQueue(this) }
    private lateinit var childName: TextView
    private lateinit var infoContainer: LinearLayout
    private lateinit var consentContainer: LinearLayout
    private lateinit var contactContainer: LinearLayout
    private lateinit var checkinButton: MaterialButton
    private var phoneNumber = ""
    private var emergencyNumber = ""
    private var parentName = ""
    private var childDisplayName = ""
    private var checkinUrl = ""
    private var userToken = ""
    private var callTarget = ""

    private val callPermissionLauncher = registerForActivityResult(ActivityResultContracts.RequestPermission()) { granted ->
        if (granted) performCall() else Toast.makeText(this, R.string.noCallPermission, Toast.LENGTH_SHORT).show()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_child_detail)
        childName = findViewById(R.id.childDetailName)
        infoContainer = findViewById(R.id.childDettailInfo)
        consentContainer = findViewById(R.id.childDetailBoolean)
        contactContainer = findViewById(R.id.detailContacts)
        checkinButton = findViewById(R.id.detailCheckin)
        checkinUrl = intent.getStringExtra(EXTRA_CHECKIN_URL).orEmpty()
        userToken = intent.getStringExtra(ChildListMainActivity.EXTRA_USER_TOKEN).orEmpty()

        findViewById<View>(R.id.detailBack).setOnClickListener { finish() }
        findViewById<View>(R.id.callParents).setOnClickListener { confirmCall(phoneNumber, parentName) }
        findViewById<View>(R.id.callEmergency).setOnClickListener { confirmCall(emergencyNumber, getString(R.string.detailEmergencyContact)) }
        checkinButton.setOnClickListener { confirmCheckin() }

        val detailUrl = intent.getStringExtra(EXTRA_URL).orEmpty()
        if (detailUrl.isNotBlank() && userToken.isNotBlank()) loadDetails(detailUrl) else showError()
    }

    private fun loadDetails(detailUrl: String) {
        findViewById<View>(R.id.detailLoading).visibility = View.VISIBLE
        requestQueue.add(JsonObjectRequest(Request.Method.GET, "$detailUrl?communicationToken=$userToken", null,
            { response ->
                findViewById<View>(R.id.detailLoading).visibility = View.GONE
                if (response.optBoolean("error")) showError(response.optString("errorText")) else bindDetails(response)
            },
            { showError() },
        ))
    }

    private fun bindDetails(response: JSONObject) {
        childDisplayName = listOf(response.optString("vorname"), response.optString("name")).filter { valid(it) }.joinToString(" ")
            .ifBlank { getString(R.string.detailMissing) }
        childName.text = childDisplayName + if (response.optBoolean("hasBirthday")) "  🎈" else ""
        findViewById<TextView>(R.id.detailInitials).text = childDisplayName.split(' ').take(2).joinToString("") { it.take(1).uppercase(Locale.GERMAN) }
        val school = response.optString("schule").takeIf(::valid)
        val grade = response.optString("klasse").takeIf(::valid)?.let { "Klasse $it" }
        findViewById<TextView>(R.id.detailSubtitle).text = listOfNotNull(grade, school).joinToString(" · ").ifBlank { getString(R.string.detailMissing) }
        phoneNumber = response.optString("phone").takeIf(::valid).orEmpty()
        emergencyNumber = response.optString("emergencyPhone").takeIf(::valid).orEmpty()
        parentName = response.optString("parentsName").takeIf(::valid) ?: getString(R.string.detailParents)
        checkinUrl = response.optString("checkinUrl").takeIf(::valid) ?: checkinUrl
        checkinButton.visibility = if (checkinUrl.isBlank() || response.optBoolean("checkin")) View.GONE else View.VISIBLE
        if (response.optBoolean("checkin")) {
            findViewById<TextView>(R.id.detailStatus).apply { text = "✓ Anwesend"; setTextColor(ContextCompat.getColor(context, R.color.phoneGreen)) }
        }
        contactContainer.removeAllViews()
        addInfoRow(contactContainer, getString(R.string.detailGuardians), parentName)
        addInfoRow(contactContainer, getString(R.string.telefon), phoneNumber.ifBlank { getString(R.string.detailMissing) })
        addInfoRow(contactContainer, getString(R.string.detailEmergencyContact), emergencyNumber.ifBlank { getString(R.string.detailMissing) })
        populateInfo(response)
    }

    private fun populateInfo(response: JSONObject) {
        infoContainer.removeAllViews()
        consentContainer.removeAllViews()
        val info = response.optJSONArray("info")
        var medical: String? = null
        for (index in 0 until (info?.length() ?: 0)) {
            val entry = info?.optJSONObject(index) ?: continue
            val label = entry.optString("name").takeIf(::valid) ?: getString(R.string.detailInformation)
            val value = entry.optString("value").takeIf(::valid) ?: getString(R.string.detailMissing)
            if (label.contains("allerg", true) || label.contains("medizin", true) || label.contains("notfallset", true)) {
                medical = listOfNotNull(medical, "$label: $value").joinToString("\n")
            } else addInfoRow(infoContainer, label, value)
        }
        if (infoContainer.childCount == 0) addInfoRow(infoContainer, getString(R.string.detailInformation), getString(R.string.detailMissing))
        findViewById<MaterialCardView>(R.id.medicalCard).visibility = if (medical == null) View.GONE else View.VISIBLE
        findViewById<TextView>(R.id.medicalText).text = medical

        val booleans = response.optJSONArray("boolean")
        for (index in 0 until (booleans?.length() ?: 0)) {
            val entry = booleans?.optJSONObject(index) ?: continue
            addConsent(entry.optString("name").takeIf(::valid) ?: getString(R.string.detailInformation), entry.optBoolean("value"))
        }
        if (consentContainer.childCount == 0) addInfoRow(consentContainer, getString(R.string.detailConsents), getString(R.string.detailMissing))
    }

    private fun addInfoRow(container: LinearLayout, label: String, value: String) {
        val row = LinearLayout(this).apply { orientation = LinearLayout.HORIZONTAL; setPadding(0, dp(12), 0, dp(12)) }
        row.addView(TextView(this).apply { text = label; textSize = 13f; setTextColor(ContextCompat.getColor(context, R.color.grey)); layoutParams = LinearLayout.LayoutParams(0, -2, 1f) })
        row.addView(TextView(this).apply { text = value; textSize = 13f; gravity = android.view.Gravity.END; setTextColor(ContextCompat.getColor(context, R.color.black)); setTypeface(typeface, android.graphics.Typeface.BOLD); layoutParams = LinearLayout.LayoutParams(0, -2, 1f) })
        container.addView(row)
    }

    private fun addConsent(label: String, allowed: Boolean) {
        consentContainer.addView(TextView(this).apply {
            text = if (allowed) "✓ $label" else "✕ $label"
            textSize = 12f; setTypeface(typeface, android.graphics.Typeface.BOLD)
            setTextColor(ContextCompat.getColor(context, if (allowed) R.color.phoneGreen else R.color.statusPurple))
            setPadding(dp(10), dp(9), dp(10), dp(9))
        })
    }

    private fun confirmCheckin() {
        if (checkinUrl.isBlank() || checkinButton.isEnabled.not()) return
        AlertDialog.Builder(this).setTitle(getString(R.string.detailCheckinTitle, childDisplayName))
            .setMessage(R.string.detailCheckinMessage).setNegativeButton(R.string.detailCancel, null)
            .setPositiveButton(R.string.checkinButton) { _, _ -> executeCheckin() }.show()
    }

    private fun executeCheckin() {
        checkinButton.isEnabled = false
        checkinButton.text = getString(R.string.detailCheckingIn)
        requestQueue.add(JsonObjectRequest(Request.Method.GET, "$checkinUrl?communicationToken=$userToken", null,
            { response ->
                if (response.optBoolean("error")) restoreCheckin(response.optString("errorText")) else {
                    checkinButton.visibility = View.GONE
                    findViewById<TextView>(R.id.detailStatus).apply { text = "✓ Anwesend"; setTextColor(ContextCompat.getColor(context, R.color.phoneGreen)) }
                    Toast.makeText(this, getString(R.string.detailCheckedIn, childDisplayName), Toast.LENGTH_LONG).show()
                }
            }, { restoreCheckin() }))
    }

    private fun restoreCheckin(message: String = getString(R.string.checkinErrorFallback)) {
        checkinButton.isEnabled = true; checkinButton.text = getString(R.string.detailCheckinNow)
        Toast.makeText(this, message, Toast.LENGTH_LONG).show()
    }

    private fun confirmCall(number: String, label: String) {
        if (number.isBlank()) { Toast.makeText(this, R.string.detailPhoneMissing, Toast.LENGTH_SHORT).show(); return }
        callTarget = number
        AlertDialog.Builder(this).setTitle(label).setMessage(getString(R.string.detailCallMessage, number))
            .setPositiveButton(R.string.detailCall) { _, _ -> requestCallPermissionOrCall() }.setNegativeButton(R.string.detailCancel, null).show()
    }

    private fun requestCallPermissionOrCall() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CALL_PHONE) == PackageManager.PERMISSION_GRANTED) performCall()
        else callPermissionLauncher.launch(Manifest.permission.CALL_PHONE)
    }

    private fun performCall() = startActivity(Intent(Intent.ACTION_CALL, Uri.parse("tel:$callTarget")))
    private fun showError(message: String = getString(R.string.detailLoadError)) { findViewById<View>(R.id.detailLoading).visibility = View.GONE; Toast.makeText(this, message, Toast.LENGTH_LONG).show() }
    private fun valid(value: String) = value.isNotBlank() && value != "null"
    private fun dp(value: Int) = (value * resources.displayMetrics.density).toInt()

    companion object { const val EXTRA_URL = "detailUrl"; const val EXTRA_CHECKIN_URL = "checkinUrl" }
}
